### PR TITLE
Security Reports

### PR DESCRIPTION
**Pull request by smallinsky**
## What 

Audit Report implementation: 
* Added AuditQuery, Report, ReprotState backend resource. 
* Added Proxy, Auth reports types cache implementing. 
* AuditQuery and Report run audit events. 
* Audit report grpc client. 
* Added HasMFA session field to KubeRequest audit event. 
* Added token expiration time to JoinInstance audit event. 

If that will help I can in Review I can split this PR into smaller chunks. 

Related: https://github.com/gravitational/teleport.e/pull/2038

### CODE REVIEW

File: lib/auth/helpers.go [hunk 1]
```diff
@@ -754,7 +754,7 @@ func NewTestTLSServer(cfg TestTLSServerConfig) (*TestTLSServer, error) {
 		return nil, trace.Wrap(err)
 	}
 
-	srv.TLSServer, err = NewTLSServer(TLSServerConfig{
+	srv.TLSServer, err = NewTLSServer(nil, TLSServerConfig{
```
Comment (r0mant) [inline lib/auth/helpers.go:757]: I agree with linter here. Can we pass context.TODO() here at least?

File: lib/services/local/secreports.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	return r, nil
+}
+
+// ListSecurityAuditQueries returns a list of audit queries.
+func (s *SecReportsService) ListSecurityAuditQueries(ctx context.Context, pageSize int, nextToken string) ([]*secreports.AuditQuery, string, error) {
+	items, nextToken, err := s.auditQuerySvc.ListResources(ctx, pageSize, nextToken)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return items, nextToken, err
```
Comment (r0mant) [inline lib/services/local/secreports.go:131]:
```suggestion
	return items, nextToken, nil
```

File: lib/services/local/secreports.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	return r, nil
+}
+
+// ListSecurityReports returns a list of security reports.
+func (s *SecReportsService) ListSecurityReports(ctx context.Context, i int, token string) ([]*secreports.Report, string, error) {
+	items, nextToken, err := s.securityReportSvc.ListResources(ctx, i, token)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return items, nextToken, err
```
Comment (r0mant) [inline lib/services/local/secreports.go:162]:
```suggestion
	return items, nextToken, nil
```

File: tool/tctl/common/resource_command.go [hunk 4]
```diff
@@ -1308,6 +1311,17 @@ func (rc *ResourceCommand) Delete(ctx context.Context, client auth.ClientI) (err
 			return trace.Wrap(err)
 		}
 		fmt.Printf("Access list %q has been deleted\n", rc.ref.Name)
+	case types.KindAuditQuery:
+		if err := client.SecReportsClient().DeleteSecurityAuditQuery(ctx, rc.ref.Name); err != nil {
+			return trace.Wrap(err)
+		}
+		fmt.Printf("Audit Query %q has been deleted\n", rc.ref.Name)
```
Comment (r0mant) [inline tool/tctl/common/resource_command.go:1318]:
```suggestion
		fmt.Printf("Audit query %q has been deleted\n", rc.ref.Name)
```

File: tool/tctl/common/resource_command.go [hunk 5]
```diff
@@ -1308,6 +1311,17 @@ func (rc *ResourceCommand) Delete(ctx context.Context, client auth.ClientI) (err
...
+	case types.KindSecurityReport:
+		if err := client.SecReportsClient().DeleteSecurityReport(ctx, rc.ref.Name); err != nil {
+			return trace.Wrap(err)
+		}
+		fmt.Printf("Security Report %q has been deleted\n", rc.ref.Name)
```
Comment (r0mant) [inline tool/tctl/common/resource_command.go:1323]:
```suggestion
		fmt.Printf("Security report %q has been deleted\n", rc.ref.Name)
```

File: gen/go/eventschema/getters.go [hunk 6]
```diff
@@ -71,87 +71,132 @@ func QueryableEventList() (string, error) {
...
-		line, err := prop.TableSchema([]string{prop.Name})
-		if err != nil {
-			return "", trace.Wrap(err)
-		}
-		sb.WriteString(line)
+	sb.WriteString("SELECT\n")
+	sb.WriteString("  event_date, event_time\n")
+	for _, v := range d.Columns {
+		sb.WriteString(viewSchemaLine(v.NameJSON(), v.NameSQL(), v.Type))
 	}
```
Comment (r0mant) [inline gen/go/eventschema/getters.go:92]: Is this safe w/r/t SQL injections? Can any part of this be user-provider? I would add a comment explaining why using string builder to assemble a query is safe here.

Same for all other places where string builder is used.

**Review by r0mant — APPROVED** [review]
Looks like this is mostly a bunch of our standard boilerplate so don't have a whole lot of feedback.

**Review by avatus — APPROVED** [review]

File: gen/go/eventschema/getters.go (hunk 6 above)
Comment (smallinsky) [reply gen/go/eventschema/getters.go:95]: This was exampled in https://github.com/gravitational/teleport/pull/33459/files/9f11cf6028aa27ce77be6381d2728882cc7f84c6#diff-088c806be4a3b1cd50cb8c07a9cffdda170738e3f746a12863f27aa86d32586b)

Namely Access Monitoring feature allows to run any arbitrary SQL query by a user where the Table permissions are handled by IAM.

File: api/proto/teleport/secreports/v1/secreports_service.proto [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+import "google/protobuf/empty.proto";
+import "teleport/header/v1/resourceheader.proto";
+import "teleport/secreports/v1/secreports.proto";
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/secreports/v1;secreportsv1";
+
+// SecReportsService is a service that manages security reports.
+service SecReportsService {
+  // UpsertAuditQuery upsets an audit query.
+  rpc UpsertAuditQuery(UpsertAuditQueryRequest) returns (google.protobuf.Empty);
```
Comment (espadolini) [inline api/proto/teleport/secreports/v1/secreports_service.proto:28]: We should really be returning the resource on any write; by returning a pbempty we have no way to return the revision, and anything that depends on knowing the value that was written (like terraform) needs a second roundtrip to read the resource again, with no real way to know if it has changed in the meantime.

Comment (smallinsky) [reply api/proto/teleport/secreports/v1/secreports_service.proto:28]: Good point. I will migrate to Upsert call where old value is returned. Thanks.

**Pull request merged**